pkg/image/admission/admission.go

```diff
@@ -0,0 +1,302 @@
+	computer := imagequota.NewGenericImageStreamUsageComputer(a.osClient, true, false)
+
+	previousImages := map[string]imageapi.Image{}
+	if previousIS != nil {
+		// get images referenced in the previous image stream so we can filter them out of
+		// the new images stream so we are only guarding against new images
+		handler := newImageRecordingHandler(previousImages)
+		err := computer.ProcessImageStreamImages(previousIS, handler)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	images := map[string]imageapi.Image{}
+	handler := newImageRecordingHandler(images)
+	err := computer.ProcessImageStreamImages(is, handler)
+	if err != nil {
+		return nil, err
+	}
+
+	// filter out anything in previous
+	imagesToProcess := map[string]imageapi.Image{}
+	for _, img := range images {
+		if _, ok := previousImages[img.Name]; !ok {
+			imagesToProcess[img.Name] = img
+		}
+	}
+
+	glog.V(4).Infof("%s: prevImages: %v, newImages: %v, toProcess: %v", PluginName, len(previousImages), len(images), len(imagesToProcess))
+
+	return imagesToProcess, nil
+}
+
+// SetOpenshiftClient provides init functionality for anything that requires the OpenShift client.
+func (a *imageLimitRangerPlugin) SetOpenshiftClient(c client.Interface) {
+	a.osClient = c
+
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return a.osClient.ImageStreams(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return a.osClient.ImageStreams(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&imageapi.ImageStream{},
+		a.imageStreamStore,
+		0,
+	)
+	a.imageStreamReflector = reflector
+
+	a.Run()
+}
+
+// Validate ensures that the OpenShift client is set on the plugin.
+func (a *imageLimitRangerPlugin) Validate() error {
+	if a.osClient == nil {
+		return fmt.Errorf("%s needs an Openshift client", PluginName)
+	}
+	return nil
+}
+
+// AdmitImage checks if the size is greater than the limit range.  Abstracted for reuse in the registry.
+func AdmitImage(size int64, limit kapi.LimitRangeItem) error {
+	imageQuantity := resource.NewQuantity(size, resource.BinarySI)
+
+	if limit.Type == imageapi.LimitTypeImageSize {
+		if limitQuantity, ok := limit.Max[kapi.ResourceStorage]; ok {
+			if limitQuantity.Cmp(*imageQuantity) < 0 {
+				// image size is larger than the permitted limit range max size, image is forbidden
+				return fmt.Errorf("%s exceeds the maximum %s usage per %s (%s)", imageQuantity.String(), kapi.ResourceStorage, imageapi.LimitTypeImageSize, limitQuantity.String())
+			}
+		}
+	}
+	return nil
+}
+
+// newImageRecordingHandler provides a handler that will record each image seen in images.  Images
+// will be keyed by image.Name.
+func newImageRecordingHandler(images map[string]imageapi.Image) imagequota.ImageStreamComputerFunc {
```

Comment:
Why is this a map of string -> image instead of string -> `*Image`?  Doing a lot more copies than you need.